cirq/contrib/paulistring/clifford_optimize.py

```diff
@@ -0,0 +1,177 @@
+# Copyright 2018 The ops Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Tuple, cast
+
+from cirq import ops, circuits
+from cirq.contrib.paulistring.pauli_string_phasor import (
+    PauliStringPhasor)
+from cirq.contrib.paulistring.convert_gate_set import (
+    converted_gate_set)
+
+
+def clifford_optimized_circuit(circuit: circuits.Circuit,
+                               tolerance: float = 1e-8
+                               ) -> circuits.Circuit:
+    # Convert to a circuit with CliffordGates, CZs and other ignored gates
+    c_cliff = converted_gate_set(circuit, no_clifford_gates=False,
+                                 tolerance=tolerance)
+
+    all_ops = list(c_cliff.all_operations())
+
+    def find_merge_point(start_i: int,
+                         string_op: PauliStringPhasor,
+                         stop_at_cz: bool,
+                         ) -> Tuple[int, PauliStringPhasor]:
+        def continue_condition(op: ops.Operation,
+                               current_string: PauliStringPhasor,
+                               is_first: bool) -> int:
+            if (isinstance(op, ops.GateOperation)
+                and isinstance(op.gate, ops.CliffordGate)):
+                return len(current_string.pauli_string) != 1
+            if (isinstance(op, ops.GateOperation)
+                and isinstance(op.gate, ops.Rot11Gate)):
+                return not stop_at_cz
+            if (isinstance(op, PauliStringPhasor)
+                and len(op.qubits) == 1
+                and (op.pauli_string[op.qubits[0]]
+                     == current_string.pauli_string[op.qubits[0]])):
+                return 2
+            return False
```

Comment:
This passes typecheck? False isn't an int..

Reply:
```
>>> False + False
0
```

Reply:
```
>>> +False
0
```